lib/puppet/functions/hiera_vault.rb

```diff
@@ -263,11 +263,12 @@ def vault_get(key, options, context)
               secret = version == :v1 ? response.data : response.data[:data]
             rescue Vault::HTTPConnectionError
               msg = "[hiera-vault] Could not connect to read secret: #{secretpath}"
+              context.explain { msg }
               raise Puppet::DataBinding::LookupError, msg
             rescue Vault::HTTPError => e
               msg = "[hiera-vault] Could not read secret #{secretpath}: #{e.errors.join("\n").rstrip}"
-              raise Puppet::DataBinding::LookupError, msg if strict_mode or e.code != 403
```

Comment:
you removed the `or e.code != 403`; this was relevant in my PR, so curious why you removed it ...?

Reply:
Oh yeah that was the other part I was going to mention: I didn't want this to be a breaking change for anyone using the lookup already. This would now error out and kill any puppet runs if there's a 403.

You can do hiera lookups with defaults if the value is not found, which is why the errors aren't raised as part of the lookup. If people want to opt in to having a full puppet failed run if there's issues with conneting to Vault, then they can use the new strict_mode option

Make sense?

Reply:
But then they get empty/default secrets, and not the secrets they expected? I'd rather have a (completely) failed Puppet run over a happy Puppet but all my secrets set to defaults......

Reply:
in fact, and as an aside, I have this snippet in my puppet manifests since the beginning of Vault-usage:

```puppet
$variable = lookup('vault_vault-test', String, 'first', 'vault-is-not-working')

case $variable {
  'vault-is-not-working': {
    fail('Vault is definitely NOT working')
  }
  'vault-is-working': {
    notice('Vault is definitely working')
  }
  default: {
    fail("Vault is behaving strangely; got this string: '${variable}'")
  }
}
```

This will make sure Puppet fails because the value is written in Vault and nowhere else.

Reply:
My knowledge might be out of date (I haven't worked at Puppet since 2017 and haven't worked in anger with a large Puppet environment since then) I think for me it's the difference between an explict lookup in code, and [Automatic Parameter Lookup within Hiera](https://puppet.com/docs/puppet/7/hiera_automatic.html#class_parameters). Especially when doing bootstrapping when the Vault servers not fully setup yet. 

Regardless, I'd rather put in your feature as an opt-in, then maybe change the 403 to error with an opt_out config in a future major change, to preserve Semver. Is that ok?

Reply:
Of course this is OK, it's your code :smile: 

We use automatic parameter lookup in hiera for everything related to vault, including the above snippet, but this may fail and therefore give us nil values. That's what I don't like to have when I expect actual values.

The case we had, was the token used by puppet to do the lookups had failed to refresh, and therefore returned empty strings as a result, which were set in database client configurations (as password), which then triggered "too many failed auths", which then brought down the whole application. It was ugly.

Reply:
Cool, gonna put this in as a non-breaking change then flip the switch in the next SemVer release 👍🏻